drivers/rpmsg/rpmsg_syslog.c

```diff
@@ -0,0 +1,157 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Copyright (C) 2017 Pinecone Inc.
+ */
+
+#include <linux/module.h>
+#include <linux/rpmsg.h>
+#include <linux/slab.h>
+
+#define RPMSG_SYSLOG_TRANSFER		0
+#define RPMSG_SYSLOG_TRANSFER_DONE	1
+#define RPMSG_SYSLOG_SUSPEND		2
+#define RPMSG_SYSLOG_RESUME		3
+
+struct rpmsg_syslog_header {
+	uint32_t			command;
+	int32_t				result;
+} __packed;
+
+struct rpmsg_syslog_transfer {
+	struct rpmsg_syslog_header	header;
+	uint32_t			count;
+	char				data[0];
+} __packed;
+
+#define rpmsg_syslog_suspend		rpmsg_syslog_header
+#define rpmsg_syslog_resume		rpmsg_syslog_header
+#define rpmsg_syslog_transfer_done	rpmsg_syslog_header
+
+struct rpmsg_syslog {
+	char				*tmpbuf;
+	unsigned int			nextpos;
+	unsigned int			alloced;
+};
+
+static int rpmsg_syslog_callback(struct rpmsg_device *rpdev,
+		void *data, int len, void *priv_, u32 src)
+{
+	struct rpmsg_syslog *priv = dev_get_drvdata(&rpdev->dev);
+	struct rpmsg_syslog_header *header = data;
+
+	if (header->command == RPMSG_SYSLOG_TRANSFER) {
+		struct rpmsg_syslog_transfer *msg = data;
+		struct rpmsg_syslog_transfer_done done;
+		unsigned int copied = msg->count;
+		unsigned int printed = 0;
+		const char *nl;
+
+		nl = memrchr(msg->data, '\n', msg->count);
+		if (nl != NULL) {
+			printed = nl + 1 - msg->data;
+			copied = msg->count - printed;
+
+			if (priv->nextpos) {
+				pr_info(KERN_TIME "%.*s%.*s", priv->nextpos,
```

Comment:
dev_info or dev_dbg ?

Reply:
Since the remote side may already add some prefix(e.g. dsp: or sensor: ), call pr_info instead of dev_info could avoid the prefix flooding.